platform-hub-api/app/services/agents/keycloak_agent_service.rb

```diff
@@ -0,0 +1,141 @@
+module Agents
+  class KeycloakAgentService
+
+    module Errors
+      class KeycloakIdentityMissing < StandardError; end
+      class KeycloakUserRepresentationMissing < StandardError; end
+      class KeycloakAccessTokenRequestFailed < StandardError; end
+      class KeycloakAccessTokenExpired < StandardError; end
+    end
+
+    attr_accessor :token, :token_expires_at
+
+    def initialize base_url:, realm:, client_id:, client_secret:, username:, password:
+      @base_url = base_url
+      @realm = realm
+      @client_id = client_id
+      @client_secret = client_secret
+      @username = username
+      @password = password
+    end
+
+    def client
+      @client ||= Faraday.new @base_url
+    end
+
+    def bearer_token
+      return token if token.present? && token_not_expired?
+
+      begin
+        response = client.post "/auth/realms/#{@realm}/protocol/openid-connect/token",
+          {
+            client_id: @client_id,
+            client_secret: @client_secret,
+            username: @username,
+            password: @password,
+            grant_type: 'password'
+          }
+      rescue => e
+        Rails.logger.error "Couldn't obtain access token from Keycloak: #{e.class} - #{e.message}"
+        raise
+      end
+
+      if response.success?
+        Rails.logger.info 'Obtained access token from Keycloak.'
+        @token = JSON.parse(response.body)['access_token']
+        @token_expires_at = JSON.parse(Base64.decode64(@token.split('.').second))['exp']
+        @token
+      else
+        Rails.logger.error "Couldn't obtain access token from Keycloak. Response: #{response.body}"
+        raise Errors::KeycloakAccessTokenRequestFailed
+      end
+    end
+
+    def get_user_representation(user)
+      keycloak_identity = user.identity(:keycloak)
+      user_id = keycloak_identity.try(:user_id)
+
+      raise Errors::KeycloakIdentityMissing if user_id.nil?
+
+      begin
+        response = client.get "/auth/admin/realms/#{@realm}/users/#{user_id}",
+          {},
+          {Authorization: "Bearer #{bearer_token}"}
+      rescue => e
+        Rails.logger.error "Couldn't get user representation from Keycloak: #{e.class} - #{e.message}"
```

Comment:
It's very likely that our User IDs will be outdated (because of things like realm consolidation, etc.) Which is why [we recently changed the auth and user lookup to rely on the email address](https://github.com/UKHomeOffice/platform-hub/pull/59). I think we have two options here:

1. Make the ID more reliable and up to date, but updating the value whenever encounter a new one (this would be done during the auth check and user lookup), or
1. Make a query to Keycloak using the email address and not the ID.

What do you think?